lib/ansible/modules/cloud/azure/azure_rm_sqlserver_facts.py

```diff
@@ -0,0 +1,195 @@
+  - name: Get instance of SQL Server
+    azure_rm_sqlserver_facts:
+      resource_group: resource_group_name
+      server_name: server_name
+
+  - name: List instances of SQL Server
+    azure_rm_sqlserver_facts:
+      resource_group: resource_group_name
+'''
+
+RETURN = '''
+id:
+    description:
+        - Resource ID.
+    returned: always
+    type: str
+    sample: /subscriptions/00000000-1111-2222-3333-444444444444/resourceGroups/sqlcrudtest-7398/providers/Microsoft.Sql/servers/sqlcrudtest-4645
+name:
+    description:
+        - Resource name.
+    returned: always
+    type: str
+    sample: sqlcrudtest-4645
+type:
+    description:
+        - Resource type.
+    returned: always
+    type: str
+    sample: Microsoft.Sql/servers
+location:
+    description:
+        - Resource location.
+    returned: always
+    type: str
+    sample: japaneast
+kind:
+    description:
+        - Kind of sql server. This is metadata used for the Azure portal experience.
+    returned: always
+    type: str
+    sample: v12.0
+version:
+    description:
+        - The version of the server.
+    returned: always
+    type: str
+    sample: 12.0
+state:
+    description:
+        - The state of the server.
+    returned: always
+    type: str
+    sample: Ready
+fully_qualified_domain_name:
+    description:
+        - The fully qualified domain name of the server.
+    returned: always
+    type: str
+    sample: fully_qualified_domain_name
+'''
+
+from ansible.module_utils.azure_rm_common import AzureRMModuleBase
+
+try:
+    from msrestazure.azure_exceptions import CloudError
+    from msrestazure.azure_operation import AzureOperationPoller
+    from azure.mgmt.sql import SqlManagementClient
+    from msrest.serialization import Model
+except ImportError:
+    # This is handled in azure_rm_common
+    pass
+
+
+class AzureRMServersFacts(AzureRMModuleBase):
+    def __init__(self):
+        # define user inputs into argument
+        self.module_arg_spec = dict(
+            resource_group=dict(
+                type='str',
+                required=True
+            ),
+            server_name=dict(
+                type='str'
+            )
+        )
+        # store the results of the module operation
+        self.results = dict(
+            changed=False,
+            ansible_facts=dict()
+        )
+        self.mgmt_client = None
+        self.resource_group = None
+        self.server_name = None
+        super(AzureRMServersFacts, self).__init__(self.module_arg_spec)
+
+    def exec_module(self, **kwargs):
+        for key in self.module_arg_spec:
+            setattr(self, key, kwargs[key])
+        self.mgmt_client = self.get_mgmt_svc_client(SqlManagementClient,
+                                                    base_url=self._cloud_environment.endpoints.resource_manager)
+
+        if (self.resource_group is not None and
+                self.server_name is not None):
+            self.results['ansible_facts']['get'] = self.get()
```

Comment:
While we may have done it in the past we don't want to be adding non host based facts to the `ansible_facts` return value. It's a bit confusing as this is technically a "facts" module but because we are getting facts on something that isn't the host we are running on, this should just be returning a dict.

This is mainly because we can run this module on multiple SQL servers to get facts for different resources and we don't want to overwrite existing calls to this module. This is similar to the `stat` module where we can call stat on multiple objects and save the results in a variable using `register`

Reply:
ok, that actually makes a lot of sense. i had problem documenting it properly. i was thinking of returning a list even with single item.
on the other hand solution you propose makes a lot of sense when it comes to merging two results.

On the other hand i see some problem with documentation, as I am doing some experiments with Ansible autocomepletion for Visual Studio code. If i parse return value like this, autocompletion will think "server_name" is a fixed value. So I am wondering if there's any way to mark in the documentation that "server_name" is variable field name, and there may be more fields?

Reply:
Not that I know off unfortunately and people document it this in various different ways making this a different per module. You could not document the contains for the initial server and just provide a sample but arguably that's worse than this as people can't easily find out what value is meant to represent.